ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -21,53 +21,975 @@
+				}
+			} else {
+				break;
+			}
+		}
+		return true;
+	}
+	return false;
+}
+
+/**
+ * Evaluates a unary expression from the current index in the expression
+ * string recursively.
+ *
+ * @param[out] ret: the integer value the unary expression evaluates to
+ *
+ * @return: whether at the current location can be evaluated as a unary
+ * 			expression
+ */
+bool
+MacroParser::unaryExpression(int64_t *ret)
+{
+	int64_t operand = 0;
+	size_t bitWidth = 0;
+	bool isSigned = true;
+	if (_scanner.atSymbol(PLUS)) {
+		if (validSingleTerm(&operand)) {
+			*ret = +operand;
+			return true;
+		}
+	} else if (_scanner.atSymbol(MINUS)) {
+		if (validSingleTerm(&operand)) {
+			*ret = -operand;
+			return true;
+		}
+	} else if (_scanner.atSymbol(NOT)) {
+		if (validSingleTerm(&operand)) {
+			*ret = !operand;
+			return true;
+		}
+	} else if (_scanner.atSymbol(TILDE)) {
+		if (validSingleTerm(&operand)) {
+			*ret = ~operand;
+			return true;
+		}
+	} else if (_scanner.validTypeCast(&isSigned, &bitWidth) ) {
+		if (validSingleTerm(&operand)) {
+			switch(bitWidth) {
+			case 0:
+				*ret = (isSigned) ? (signed) operand : (unsigned) operand;
+				break;
+			case 8:
+				*ret = (isSigned) ? (int8_t) operand : (uint8_t) operand;
+				break;
+			case 16:
+				*ret = (isSigned) ? (int16_t) operand : (uint16_t) operand;
+				break;
+			case 32:
+				*ret = (isSigned) ? (int32_t) operand : (uint32_t) operand;
+				break;
+			case 64:
+				*ret = (isSigned) ? (int64_t) operand : (uint64_t) operand;
+				break;
+			default:
+				return false;
+			}
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Evaluates a parenthesized expression from the current index in the
+ * expression string recursively.
+ *
+ * @param[out] ret: the integer value the expression evaluates to
+ *
+ * @return: whether at the current location can be evaluated as a parenthesized
+ * 			expression
+ */
+bool
+MacroParser::parentheses(int64_t *ret)
+{
+	int64_t innerValue = 0;
+	if (_scanner.atSymbol(LPAREN)) {
+		if (logicalOr(&innerValue)) {
+			if (_scanner.atSymbol(RPAREN)) {
+				*ret = innerValue;
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+/**
+ * Checks if the expression at the cursor is one of: unary expression,
+ * parenthesized expression, or a number, then evaluates it.
+ *
+ * @param[out] ret: the value after evaluating the expression
+ *
+ * return: whether at the current location can be evaluated as one of the
+ * expressions listed
+ */
+bool
+MacroParser::validSingleTerm(int64_t *ret)
+{
+	return unaryExpression(ret) || parentheses(ret)	|| _scanner.readNumber(ret);
+}
+
+/*
+ * evaluates the numeric value of a constant expression from a macro
+ *
+ * @param[out] ret: the integer value of the expression in an long long
+ *
+ * return: DDR return code indicating success or error
+ */
+DDR_RC
+Macro::getNumeric(long long *ret) const
+{
+	DDR_RC rc = DDR_RC_ERROR;
+	string value = getValue();
+	int64_t retVal = 0;
+	MacroParser parser(value.c_str());
```

Comment:
There's no need to make a copy of `_value` here; the parser can be constructed from `_value.c_str()`.